gcs/builder/builder.go

```diff
@@ -0,0 +1,363 @@
+// building the filter.
+func WithRandomKey() *GCSBuilder {
+	return WithRandomKeyPN(DefaultP, 0)
+}
+
+// BuildBasicFilter builds a basic GCS filter from a block. A basic GCS filter
+// will contain all the previous regular outpoints spent within a block, as well
+// as the data pushes within all the outputs created within a block which can be
+// spent by regular transactions.
+func BuildBasicFilter(block *wire.MsgBlock) (*gcs.Filter, error) {
+	blockHash := block.BlockHash()
+	b := WithKeyHash(&blockHash)
+
+	// Add "regular" data from stake transactions.  For each class of stake
+	// transaction, the following data is committed to the regular filter:
+	//
+	//   ticket purchases:
+	//     - all previous outpoints
+	//     - data pushes in change output scripts
+	//
+	//   votes:
+	//     - data pushes in all OP_SSGEN-tagged output scripts (all outputs
+	//       after the first two -- these describe the block voted on and the
+	//       vote choices)
+	//
+	//   revocations:
+	//     - data pushes in all output scripts
+	//
+	// Because change outputs are required in a ticket purchase, even when
+	// unused, a special case is made that excludes their commitment when the
+	// output value is zero (provably unspendable).
+	for _, tx := range block.STransactions {
+		switch stake.DetermineTxType(tx) {
+		case stake.TxTypeSStx: // Ticket purchase
+			for _, in := range tx.TxIn {
+				b.AddOutPoint(&in.PreviousOutPoint)
+			}
+			for _, out := range tx.TxOut {
+				sc := txscript.GetScriptClass(out.Version, out.PkScript)
+				if sc == txscript.StakeSubChangeTy && out.Value != 0 {
+					b.AddScript(out.PkScript)
+				}
+			}
+
+		case stake.TxTypeSSGen: // Vote
+			for _, out := range tx.TxOut {
+				sc := txscript.GetScriptClass(out.Version, out.PkScript)
+				if sc == txscript.StakeGenTy {
+					b.AddScript(out.PkScript)
+				}
+			}
+
+		case stake.TxTypeSSRtx: // Revocation
+			for _, out := range tx.TxOut {
+				b.AddScript(out.PkScript)
+			}
+		}
+	}
+
+	// For regular transactions, all previous outpoints except the coinbase's
+	// are committed, and all data pushes from output scripts (including
+	// OP_RETURN pushes) are committed.
+	for i, tx := range block.Transactions {
+		if i != 0 {
+			for _, txIn := range tx.TxIn {
+				b.AddOutPoint(&txIn.PreviousOutPoint)
+			}
+		}
+		for _, txOut := range tx.TxOut {
+			b.AddScript(txOut.PkScript)
+		}
+	}
+
+	return b.Build()
+}
+
+// BuildExtFilter builds an extended GCS filter from a block. An extended filter
+// supplements a regular basic filter by including all transaction hashes of
+// regular and stake transactions, and adding the witness data (a.k.a. the
+// signature script) found within every non-coinbase regular transaction.
+func BuildExtFilter(block *wire.MsgBlock) (*gcs.Filter, error) {
+	blockHash := block.BlockHash()
+	b := WithKeyHash(&blockHash)
+
+	// For each stake transaction, commit the transaction hash.
+	for _, tx := range block.STransactions {
```

Comment:
Based on the intent described by the comment of the function, this appears to be missing the signature scripts for all of the stake transactions.

In the case of votes and revocations, it's the output of the ticket that is being spent (`.TxIn[1]` for votes, and `.TxIn[0]` for revocations), and in the case of tickets, it is all inputs.

Reply:
The function comment says it includes the witness/sigscripts of regular transactions.  No where does it say it commits to the input scripts of stake transactions.

I'm not sure if it would be worthwhile modifying the extended filter to include this data, or if a 3rd filter type should be added specifically for extra commitments for stake transactions.